src/Rector/Property/ProphecyPHPDocRector.php

```diff
@@ -0,0 +1,184 @@
+    {
+        // PhpParser\Node\Expr\Assign
+        // PhpParser\Node\Expr\MethodCall
+        return [Assign::class];
+    }
+
+    public function refactor(Node $node)
+    {
+        if (! $this->testsNodeAnalyzer->isInTestClass($node)) {
+            return null;
+        }
+
+        if (! $node instanceof Assign) {
+            return;
+        }
+
+        $expr = $node->expr;
+
+        if (! $expr instanceof MethodCall) {
+            return;
+        }
+
+        $var = $expr->var;
+        if (! $var instanceof Variable) {
+            return;
+        }
+
+        if ($var->name !== 'this') {
+            return;
+        }
+
+        $name = $expr->name;
+        if (! $name instanceof Identifier) {
+            return;
+        }
+
+        if ($name->name !== 'prophesize') {
+            return;
+        }
+
+        $value = $expr->args[0]->value;
+
+        if ($value instanceof String_) {
+            $prophesizeClassParts = \explode('\\', $value->value);
+        } elseif ($value instanceof ClassConstFetch) {
+            $prophesizeClassParts = $value->class->parts;
+        } else {
+            return;
+        }
+
+        $var = $node->var;
+
+        if (! $var instanceof PropertyFetch) {
+            return;
+        }
+
+        $propertyName = $var->name->name;
+
+        $class = $this->betterNodeFinder->findParentType($node, Class_::class);
+
+        foreach ($class->stmts as $propertyStmt) {
+            if (! $propertyStmt instanceof Property) {
+                continue;
+            }
+
+            if ($propertyStmt->props[0]->name->name !== $propertyName) {
+                continue;
+            }
+
+            $doc = new Doc("/**\n     * @var ObjectProphecy<" . $prophesizeClassParts[array_key_last(
+                $prophesizeClassParts
+            )] . ">\n     */");
+            $propertyStmt->setDocComment($doc);
+            $propertyStmt->getDocComment();
+
+            $useClassNames = ['Prophecy\Prophecy\ObjectProphecy', \implode('\\', $prophesizeClassParts)];
+
+            $namespace = $this->betterNodeFinder->findParentType($node, Namespace_::class);
+
+            if (! $namespace instanceof Namespace_) {
+                return;
+            }
+
+            foreach ($namespace->stmts as $useStmt) {
+                if (! $useStmt instanceof Use_) {
+                    continue;
+                }
+
+                $foundClass = \implode('\\', $useStmt->uses[0]->name->parts);
+
+                $findIndex = \array_search($foundClass, $useClassNames, true);
+                if ($findIndex !== false) {
+                    unset($useClassNames[$findIndex]);
+                }
+            }
+
+            $uses = $this->nodeFactory->createUsesFromNames($useClassNames);
+            $stmts = $namespace->stmts;
+
+            foreach ($uses as $use) {
+                $stmts[] = $use; // TODO not working currently
+            }
+
+            $namespace->stmts = $stmts;
```

Comment:
this lines fails which make the current tests failing.

Reply:
By convention, we keep all class names FQN. Coding standard tool handles this, or `$rectorConfig->importNames();`.
In short, you never have to deal with FQN in any rule :)

Reply:
PHP-CS-Fixer and importNames did not work in this case, think they can not handle this kind of phpdocs.

I'm still curious what would be the correct way to add `use` statements via rector?

Reply:
In that case it should be fixed in the importNames(). This rule should not handle this.

Reply:
> I'm still curious what would be the correct way to add use statements via rector?

Check the `NameImportingPostRector`  :) that's the rule that handles it.

Reply:
Okay looks like a postprocessor is adding that I understand. Thx for the insights.

Reply:
👍